File: typescript/packages/core/src/redact/scan.ts
import { ordered } from "./registry";

// The matcher both text and bytes redaction run: the longest registered value at each position,
// then the whole output checked again, so a marker joined to its neighbours never forms a value.

/** Value → marker, in the unit `encode` gives (text, or bytes as latin1 text). */
export type Replacements = {
  readonly markers: ReadonlyMap<string, string>;
  /** What a piece of output that still holds a value becomes. */
  readonly redacted: string;
};

export function replacements(encode: (text: string) => string): Replacements {
  const markers = new Map(
    ordered().map(([v, l]) => [encode(v), encode(marker(l))]),
  );
  const values = [...markers.keys()];
  const plain = encode("[redacted]");
  return {
    markers,
    redacted: values.some((v) => plain.includes(v)) ? "" : plain,
  };
}

/** `[secret <label>]`, or `[secret]` when the label holds a registered value. */
function marker(label: string): string {
  const labelled = `[secret ${label}]`;
  return ordered().some(([v]) => labelled.includes(v)) ? "[secret]" : labelled;
}

/** Whether `text` holds any of `values` (already in the scan's unit). */
export function holdsAny(text: string, r: Replacements): boolean {
  return [...r.markers.keys()].some((v) => text.includes(v));
}

const escapeRegExp = (text: string): string =>
  text.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");

/**
 * The longest value at each position, scanning left to right. Unless `final`, it stops where
 * the rest could still grow into a longer value (`abc` before a possible `abc123`), and returns
 * that rest to hold for the next chunk.
 */
export function scan(
  pending: string,
  r: Replacements,
  final: boolean,
): { readonly out: string; readonly rest: string } {
  const values = r.markers;
  if (values.size === 0) return { out: pending, rest: "" };
  const hold = final ? pending.length : heldFrom(pending, values);
  const pattern = new RegExp(
    [...values.keys()].map(escapeRegExp).join("|"),
    "g",
  );
  let out = "";
  let at = 0;
  for (const m of pending.matchAll(pattern)) {
    if (m.index >= hold) break;
    out += pending.slice(at, m.index) + (values.get(m[0]) ?? "");
    at = m.index + m[0].length;
  }
  const cut = Math.max(hold, at);
  return { out: out + pending.slice(at, cut), rest: pending.slice(cut) };
}

/** The first position whose rest is a proper prefix of a value: it may still become one. */
function heldFrom(
  pending: string,
  values: ReadonlyMap<string, string>,
): number {
  const longest = Math.max(...[...values.keys()].map((v) => v.length));
  for (
    let i = Math.max(0, pending.length - longest + 1);
    i < pending.length;
    i += 1
  ) {
    const tail = pending.slice(i);
    if (
      [...values.keys()].some(
        (v) => v.length > tail.length && v.startsWith(tail),
      )
    )
      return i;
  }
  return pending.length;
}

/**
 * A stream redacted as it arrives. A chunk that, joined to the tail already emitted, would hold
 * a value is emitted as the plain marker instead.
 */
export function streamOf(encode: (text: string) => string): {
  readonly feed: (text: string) => string;
  readonly end: () => string;
} {
  let pending = "";
  let emitted = "";
  const take = (final: boolean): string => {
    const r = replacements(encode);
    const { out, rest } = scan(pending, r, final);
    pending = rest;
    const shown = holdsAny(emitted + out, r) ? r.redacted : out;
    // Only the last longest-1 units can join the next chunk into a value.
    const keep = Math.max(0, ...[...r.markers.keys()].map((v) => v.length - 1));
    emitted = keep === 0 ? "" : (emitted + shown).slice(-keep);
    return shown;
  };
  return {
    feed: (text) => {
      pending += text;
      return take(false);
    },
    end: () => take(true),
  };
}

File: python/src/threads/redaction/scan.py
"""The matcher text and bytes redaction both run: the longest registered value at each position,
then the whole output checked again, so a marker joined to its neighbours never forms a value."""

import re
from collections.abc import Callable
from dataclasses import dataclass

from threads.redaction.registry import ordered


@dataclass(frozen=True, slots=True)
class Replacements:
    """Value to marker, in the unit `encode` gives (text, or bytes as latin-1 text)."""

    markers: dict[str, str]
    redacted: str
    """What a piece of output that still holds a value becomes."""

    def holds(self, text: str) -> bool:
        return any(v in text for v in self.markers)


def replacements(encode: Callable[[str], str]) -> Replacements:
    markers = {encode(v): encode(_marker(label)) for v, label in ordered()}
    plain = encode("[redacted]")
    return Replacements(markers, "" if any(v in plain for v in markers) else plain)


def _marker(label: str) -> str:
    """`[secret <label>]`, or `[secret]` when the label holds a registered value."""
    labelled = f"[secret {label}]"
    return "[secret]" if any(v in labelled for v, _ in ordered()) else labelled


def _held_from(pending: str, values: dict[str, str]) -> int:
    """The first position whose rest is a proper prefix of a value: it may still become one."""
    longest = max(len(v) for v in values)
    for i in range(max(0, len(pending) - longest + 1), len(pending)):
        tail = pending[i:]
        if any(len(v) > len(tail) and v.startswith(tail) for v in values):
            return i
    return len(pending)


def scan(pending: str, r: Replacements, *, final: bool) -> tuple[str, str]:
    """The longest value at each position, scanning left to right. Unless `final`, it stops
    where the rest could still grow into a longer value (`abc` before a possible `abc123`), and
    returns that rest to hold for the next chunk."""
    values = r.markers
    if not values:
        return pending, ""
    hold = len(pending) if final else _held_from(pending, values)
    pattern = re.compile("|".join(re.escape(v) for v in values))
    out: list[str] = []
    at = 0
    for m in pattern.finditer(pending):
        if m.start() >= hold:
            break
        out += [pending[at : m.start()], values[m.group(0)]]
        at = m.end()
    cut = max(hold, at)
    out.append(pending[at:cut])
    return "".join(out), pending[cut:]


class Stream:
    """A stream redacted as it arrives. A chunk that, joined to the tail already emitted, would
    hold a value is emitted as the plain marker instead."""

    def __init__(self, encode: Callable[[str], str]) -> None:
        self._encode = encode
        self._pending = ""
        self._emitted = ""

    def feed(self, text: str) -> str:
        self._pending += text
        return self._take(final=False)

    def end(self) -> str:
        return self._take(final=True)

    def _take(self, *, final: bool) -> str:
        r = replacements(self._encode)
        out, self._pending = scan(self._pending, r, final=final)
        shown = r.redacted if r.holds(self._emitted + out) else out
        # Only the last longest-1 units can join the next chunk into a value.
        keep = max((len(v) - 1 for v in r.markers), default=0)
        self._emitted = (self._emitted + shown)[-keep:] if keep else ""
        return shown
